src/app/vertextool/qgsvertextool.cpp

```diff
@@ -2135,7 +2135,7 @@ void QgsVertexTool::moveVertex( const QgsPointXY &mapPoint, const QgsPointLocato
 
   // needed to get Z value
   if ( mapPointMatch && mapPointMatch->layer() && QgsWkbTypes::hasZ( mapPointMatch->layer()->wkbType() ) && ( mapPointMatch->hasEdge() || mapPointMatch->hasMiddleSegment() ) )
-    layerPoint = mapPointMatch->interpolatedPoint();
+    layerPoint = toLayerCoordinates( dragLayer, mapPointMatch->interpolatedPoint() );
```

Comment:
here, attribute Point in the mapPointMatch object is in map coordinates, so in the interpolatedPoint() method, the closestPoint function is used with a geometry in layers coordinates and point in map coordinates. As I am not very aware about Match stuff, I don't know what is wrong: mPoint in the Match instance in map coordinates or supposing that mapPoint is in layer coordinates?

Reply:
In my latest commits I apply the same logic as for #41306

here is a full demo with my data and yours.


https://user-images.githubusercontent.com/7521540/133150942-60ac9d17-2498-4742-b883-f8ef1127033f.mp4


https://user-images.githubusercontent.com/7521540/133150979-b5dfae2f-99a4-4254-8ac7-a2bfa99b2545.mp4